![Imagen](itd.png)

# Analisis De Datos Masivos
## Práctica de laboratorio 4.1.3.6 Clasificación del árbol de decisiones
    
## Nombre Del Docente
### Jose Gabriel Rodriguez Rivas
    
## Francisco javier Favela Najera 

In [69]:
import pandas as pd
import numpy as np
import sklearn
#create a pandas dataframe called "training" from the titanic-train.csv file
training = pd.read_csv("titanic-train.csv")

Con los datos de arriba, ¿qué tipos de preguntas podemos hacer acerca de los factores que contribuyeron a que los pasajeros sobrevivan o fallecieran en la catástrofe del Titanic? el genero y el numero de cabina es importante

In [27]:
training.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 915 entries, 0 to 914
Data columns (total 12 columns):
PassengerId    915 non-null int64
Survived       915 non-null int64
Pclass         915 non-null int64
Name           915 non-null object
Gender         915 non-null object
Age            738 non-null float64
SibSp          915 non-null int64
Parch          915 non-null int64
Ticket         915 non-null object
Fare           915 non-null float64
Cabin          202 non-null object
Embarked       914 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 85.9+ KB


¿Faltan valores en el conjunto de datos? faltan valores de cabina y de edad 

In [28]:
training.head()

PassengerId  Survived  Pclass                         Name  Gender   Age  \
0            1         0       1       Davidson, Mr. Thornton    male  31.0   
1            2         0       3              Asim, Mr. Adola    male  35.0   
2            3         0       3           Nankoff, Mr. Minko    male   NaN   
3            4         0       1     Thayer, Mr. John Borland    male  49.0   
4            5         0       3  Strandberg, Miss. Ida Sofia  female  22.0   

   SibSp  Parch              Ticket      Fare Cabin Embarked  
0      1      0          F.C. 12750   52.0000   B71        S  
1      0      0  SOTON/O.Q. 3101310    7.0500   NaN        S  
2      0      0              349218    7.8958   NaN        S  
3      1      1               17421  110.8833   C68        C  
4      0      0                7553    9.8375   NaN        S

## Paso 2: Preparar los datos para el modelo de árbol de decisiones

In [29]:
training["Gender"] = training["Gender"].apply(lambda toLabel: 0 if toLabel == 'male' else 1)

In [30]:
training.head()

PassengerId  Survived  Pclass                         Name  Gender   Age  \
0            1         0       1       Davidson, Mr. Thornton       0  31.0   
1            2         0       3              Asim, Mr. Adola       0  35.0   
2            3         0       3           Nankoff, Mr. Minko       0   NaN   
3            4         0       1     Thayer, Mr. John Borland       0  49.0   
4            5         0       3  Strandberg, Miss. Ida Sofia       1  22.0   

   SibSp  Parch              Ticket      Fare Cabin Embarked  
0      1      0          F.C. 12750   52.0000   B71        S  
1      0      0  SOTON/O.Q. 3101310    7.0500   NaN        S  
2      0      0              349218    7.8958   NaN        S  
3      1      1               17421  110.8833   C68        C  
4      0      0                7553    9.8375   NaN        S

In [31]:
training["Age"].fillna(training["Age"].mean(), inplace=True)

In [32]:
training.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 915 entries, 0 to 914
Data columns (total 12 columns):
PassengerId    915 non-null int64
Survived       915 non-null int64
Pclass         915 non-null int64
Name           915 non-null object
Gender         915 non-null int64
Age            915 non-null float64
SibSp          915 non-null int64
Parch          915 non-null int64
Ticket         915 non-null object
Fare           915 non-null float64
Cabin          202 non-null object
Embarked       914 non-null object
dtypes: float64(2), int64(6), object(4)
memory usage: 85.9+ KB


In [33]:
training.tail()

PassengerId  Survived  Pclass                        Name  Gender  \
910          911         0       3         Baccos, Mr. Raffull       0   
911          912         0       1  Smith, Mr. Richard William       0   
912          913         0       3            Wirz, Mr. Albert       0   
913          914         0       3         Johnson, Mr. Alfred       0   
914          915         0       3            Fox, Mr. Patrick       0   

           Age  SibSp  Parch  Ticket     Fare Cabin Embarked  
910  20.000000      0      0    2679   7.2250   NaN        C  
911  29.970867      0      0  113056  26.0000   A19        S  
912  27.000000      0      0  315154   8.6625   NaN        S  
913  49.000000      0      0    LINE   0.0000   NaN        S  
914  29.970867      0      0  368573   7.7500   NaN        Q

## ¿Cuál es el valor que se utilizó para reemplazar a las edades faltantes?29.970867

In [34]:
y_target = training["Survived"].values

In [35]:
columns = ["Fare", "Pclass", "Gender", "Age", "SibSp"]
#create the variable to hold the features that the classifier will use
X_input = training[list(columns)].values

In [36]:
from sklearn import tree

#create clf_train as a decision tree classifier object
clf_train = tree.DecisionTreeClassifier(criterion="entropy", max_depth=3)

#train the model using the fit() method of the decision tree object. 
#Supply the method with the input variable X_in
clf_train = clf_train.fit(X_input, y_target)

In [37]:
clf_train.score(X_input,y_target)

0.8163934426229508

In [38]:
from sklearn.externals.six import StringIO
with open("titanic.dot", 'w') as f:
  f = tree.export_graphviz(clf_train, out_file=f, feature_names=columns)

In [39]:
dot -Tpng .titanic.dot -o .titanic.png

NameError: name 'dot' is not defined

In [40]:
from IPython.display import Image

#display the decison tree graphic
Image("titanic.png")

TypeError: a bytes-like object is required, not 'str'

TypeError: a bytes-like object is required, not 'str'

<IPython.core.display.Image object>

## ¿Qué describe al grupo que tenía mayor cantidad de víctimas fatales? ¿Qué grupo tenía la mayor cantidad de sobrevivientes?

## Parte 2: Aplicar el modelo del árbol de decisiones

In [41]:
testing = pd.read_csv("titanic-test.csv")
testing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 393 entries, 0 to 392
Data columns (total 12 columns):
PassengerId    393 non-null int64
Survived       393 non-null int64
Pclass         393 non-null int64
Name           393 non-null object
Gender         393 non-null object
Age            307 non-null float64
SibSp          393 non-null int64
Parch          393 non-null int64
Ticket         393 non-null object
Fare           393 non-null float64
Cabin          93 non-null object
Embarked       392 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 36.9+ KB


## ¿Cuántos registros hay en el conjunto de datos? 393

## ¿Qué variable(s) importante(s) son valores faltantes y cuántas faltan? Cabin Y Age,301 y 86

In [42]:
testing["Gender"] = testing["Gender"].apply(lambda toLabel: 0 if toLabel == 'male' else 1)

In [43]:
testing["Age"].fillna(testing["Age"].mean(), inplace=True)

In [44]:
testing.head()

PassengerId  Survived  Pclass  \
0          916         0       2   
1          917         1       1   
2          918         0       3   
3          919         1       1   
4          920         1       3   

                                                Name  Gender   Age  SibSp  \
0                    Coleridge, Mr. Reginald Charles       0  29.0      0   
1  Spedden, Mrs. Frederic Oakley (Margaretta Corn...       1  40.0      1   
2                                Windelov, Mr. Einar       0  21.0      0   
3                             Minahan, Miss. Daisy E       1  33.0      1   
4                   Wilkes, Mrs. James (Ellen Needs)       1  47.0      1   

   Parch            Ticket    Fare Cabin Embarked  
0      0       W./C. 14263   10.50   NaN        S  
1      1             16966  134.50   E34        C  
2      0  SOTON/OQ 3101317    7.25   NaN        S  
3      0             19928   90.00   C78        Q  
4      0            363272    7.00   NaN        S

In [47]:
X_input = testing[list(columns)].values

In [48]:
target_labels = clf_train.predict(X_input)
target_labels = pd.DataFrame({'Est_Survival':target_labels, 'Name':testing['Name']})

In [57]:
import numpy as np
# Load data for all passengers in the variable all_data
all_data = pd.read_csv("titanic_all.csv")
# Merging using the field Name as key, selects only the rows of the two datasets that refer to the same passenger
testing_results = pd.merge(target_labels, all_data[['Name','Survived']], on=['Name'])

# Compute the accuracy as a ratio of matching observations to total osbervations. Store this in in the variable acc.
acc = np.sum(testing_results['Est_Survival'] == testing_results['Survived']) / float(len(testing_results))
print(acc)

0.7588832487309645


## Parte 3: Evaluar el modelo del árbol de decisiones

In [59]:
all_data = pd.read_csv("titanic_all.csv", usecols=['Survived','Pclass','Gender','Age','SibSp','Fare'])
all_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1308 entries, 0 to 1307
Data columns (total 6 columns):
Survived    1308 non-null int64
Pclass      1308 non-null int64
Gender      1308 non-null object
Age         1045 non-null float64
SibSp       1308 non-null int64
Fare        1308 non-null float64
dtypes: float64(2), int64(3), object(1)
memory usage: 61.4+ KB


## ¿Cuántos registros hay en el conjunto de datos? 1308

## ¿Qué variable(s) importante(s) son valores faltantes y cuántas faltan? Age y 263

In [64]:
all_data["Gender"] = all_data["Gender"].apply(lambda toLabel: 0 if toLabel == 'male' else 1)

In [67]:
all_data["Age"].fillna(all_data["Age"].mean(), inplace=True)

In [72]:
#create the input and target variables as uppercase X and lowercase y. Reuse the columns variable.
X = all_data[list(columns)].values
y = all_data["Survived"].values

#generate the four testing and training data arrays with the train_test_split() method
X_train,X_test,y_train,y_test=train_test_split(X, y, test_size=0.40, random_state=0)

In [74]:
type(X_train)

numpy.ndarray

In [75]:
clf_train = tree.DecisionTreeClassifier(criterion="entropy", max_depth=3)

#fit the training model using the input and target variables
clf_train = clf_train.fit(X_train, y_train)

In [76]:
train_score = str(clf_train.score(X_train,y_train))
test_score = str(clf_train.score(X_test,y_test))

#output the values in a test string
print('Training score = '+ train_score+' Testing score = '+test_score)

Training score = 0.7104591836734694 Testing score = 0.6851145038167938
